### ISSUE TITLE
L<foo|#method bar> Links Are Broken on Individual Generated Pages

### ISSUE THREAD

**Issue by zoffixznet** [perl6/doc#162]
If you visit http://doc.perl6.org/routine/push and try to click the `method append` link at the bottom of `routine push` section, nothing will happen, because target `#method_push` doesn't exist on that page. It does exist on the full page, from which the aforementioned page was generated: http://doc.perl6.org/type/Array#method_append

Whatever's generating those pages need to account that `L<>` links need to be expanded. Currently, all of these links are likely broken (grep on `doc/` dir of this repo:)

``` bash
$ grep -iER '[<|]#method' .
./Type/CurrentThreadScheduler.pod:that L<#method cue> blocks until the code has finished executing.
./Type/IO/Socket.pod:connection. The binary version is L<#method write>.
./Type/IO/Socket.pod:connection. The string version is L<#method send>.
./Type/IO/Socket/INET.pod:For text operations (such as L<#method lines> and L<#method get>),
./Type/Supply.pod:The L<live method|#method live> returns C<True> on live supplies. Factory
./Type/Supply.pod:methods such as L<interval|#method interval>, L<from-list|#method from-list>
./Type/Metamodel/AttributeContainer.pod:Returns a true value if L<method set_rw|#method set_rw> has been called on
./Type/Array.pod:See L<method append|#method append> for when you want to append multiple
./Type/Array.pod:The notes in L<the documentation for method push|#method push> apply,
./Type/Array.pod:L<#method prepend> is the equivalent for adding multiple elements from one
./Type/Match.pod:Returns the payload that was set with L<#method make>.
./Type/Match.pod:Alias for L<#method made>.
./Type/Proc/Async.pod:You must call C<stdout> before you call L<#method start>. Otherwise an
./Type/Proc/Async.pod:You must call C<stderr> before you call L<#method start>. Otherwise an
./Type/Proc/Async.pod:waiting for the promise from C<#method start> hangs (for a program opened for
./Type/Cool.pod:The inverse operation is L<chr|#method chr>.
./Type/Cool.pod:The inverse operation is L<ord|#method ord>.
./Type/DateTime.pod:L<#method timezone>.
./Type/DateTime.pod:The formatting function is called by L<#method Str> with the invocant as its
./Type/DateTime.pod:L<the formatter|#method formatter>.
./Type/DateTime.pod:Negative offsets are allowed, though L<#method earlier> is more idiomatic for
./Type/DateTime.pod:towards the past applied. See L<#method later> for usage.
./Type/Parameter.pod:L<alias|#method named_names>.  Often, the name will chosen descriptively
./Type/Parameter.pod:L<rw|#method_rw>-parameter which can only bind to variables, never to values.
./Type/Parameter.pod:L<sigil|#method sigil> of C<'\'>, which is not really a sigil insofar
./Language/subscripts.pod:See L<#method BIND-POS> and L<#method BIND-KEY> for the underlying mechanism.
./Language/subscripts.pod:See L<#method EXISTS-POS> and L<#method EXISTS-KEY> for the underlying
./Language/subscripts.pod:See L<#method DELETE-POS> and L<#method DELETE-KEY> for the underlying
./Language/subscripts.pod:|#Methods_to_implement_for_positional_subscripting>) for your custom type.
```

**Comment by paultcochrane** [comment]
Yes, this is a problem which I keep noticing.  AFAIK most previous attempts to fix isolated cases link to the relevant page with the full path in the link.  Would full paths in the links possibly be a better solution?

**Comment by zoffixznet** [comment]
IMHO that would not be the full solution. Someone eventually will type the short version and we'll have broken links again.

If the program that generates the pages watches for any `L<#foo>` or `L<foo|#foo>` and automatically generates the full proper link, then we won't see this issue again, even if someone uses the short version :smile_cat:

**Comment by ShimmerFairy** [comment]
I think one possible solution would be if the `doc:` scheme was better at accessing specific things (I don't know how much `Pod::To::HTML` or the htmlification does at the moment). I'm thinking that things like `L<doc:Array#method push>` should Do The Right Thing™. This doesn't fix the 

Another solution would be that the documentation isn't split up in the first place. I'm guessing the reason for it is to make finding info on routines easier, when they can exist on different objects. I think for that I'd prefer wiki-style disambiguation pages (disambiguating on the possible objects) over the current practice of copy-pasting bits of other files and assembling them.

As an example, the page for `push` would mention the possibilities of `Array` and `Hash`, linking to the appropriate sections. For disambiguations where it may not be easy or immediately obvious for a user to know which object they mean, it should of course offer some explanation on figuring it out.

Point is, I can't help but feel that the current practice forces you to write sections for at least two documents, the actual document you're writing in and any "virtual" ones. It'd be a lot nicer in my view if we didn't do that.

(By the way, I just went looking quickly to see the examples ptc mentioned, and some of those "full link" solutions *should* be parse-time failures, e.g. `L<classes|/language/objects#Classes>` . The only time you can eliminate the scheme is when your link looks like `L<doc:#foo>` (that is, referring to a section in the current document), otherwise the `doc:` or `http:` has to be there.)

**Comment by AlexDaniel** [comment]
Almost one year later, this is still relevant and has to be fixed. Also, there are now twice more lines found by the provided grep command.

**Comment by JJ** [comment]
Related also to #1761 and #1542, as well as #1433

**Comment by coke** [comment]
This has nothing to do with 1761.

**Comment by JJ** [comment]
@coke removed.

**Comment by JJ** [comment]
Append is now linked to https://docs.perl6.org/routine/append. However, the problem with the rest of the links is that they have no underscore, basically. I haven't checked them one by one, but some of them have been edited. So I wonder if this would better be closed and absorbed into #561

**Comment by JJ** [comment]
This has apparently been fixed programmatically somewhere. Even if it continues in that way in the documents, underscores have been added. So closing this.
